Add support to cuDNN CTC loss

kaixih opened the pull request:
- This PR supports CUDNN CTC Loss as the backend of ctc_loss_v2()
- Users need to use the environment variable TF_CUDNN_CTC_LOSS=1

- Why can we make it default (not using TF_CUDNN_CTC_LOSS)?
   - ctc_loss_v2 supports a variable blank index but will transpose it to the last index before calling the actual implementation. However, CUDNN implementation only supports the 0th blank index. This indicates ctc_loss_v2 has to select the correct implementation based on platform, which cannot be done inside operation definition. 

- What is the logic in the new ctc_loss_v2()?
   - _ctc_loss_impl() will call the actual implementation based on given use_cudnn parameter
      - If true, call the CUDNN implementation
      - If false, call the original implementation 
   - ctc_loss_v2() will transpose the blank index to 0 if TF_CUDNN_CTC_LOSS=1 and then call _ctc_loss_impl(use_cudnn=true)
   - ctc_loss_v2() will transpose the blank index to the last index if TF_CUDNN_CTC_LOSS=0 and then call _ctc_loss_impl(use_cudnn=false)

fyi @nluehr

kaixih committed:
Add changes to support cuDNN CTC loss

kaixih committed:
CPU CTC tests without V2 and update goldens

kaixih committed:
Added pbtxt for ctc loss v2

kaixih committed:
Changed some positions of macros for cuDNN CTC loss

kaixih committed:
Switch to non-deterministic algo which allow larger label size

aaroey commented:
@chsigg could you help to take a look at this? Thanks

rmlarsen commented on tensorflow/core/kernels/ctc_loss_op.cc:
```diff
@@ -25,8 +29,86 @@ limitations under the License.
+                 std::vector<int> *labels_lengths) {
+  const T* h_in = labels_indices->flat<T>().data();
+  for(int i = 0; i < num_indices; i++) {
+    T key = h_in[i * 2];
```
const ref

rmlarsen commented on tensorflow/core/kernels/ctc_loss_op.cc:
```diff
@@ -25,8 +29,86 @@ limitations under the License.
+// takes the ownership of the underlying memory. The expectation is that the
+// memory should be alive for the span of the cudnnCTCLoss itself.
+template <typename T>
+class CudnnCtcLossAllocatorInTemp : public ScratchAllocator {
```
This code is identical to e.g. CudnnBatchNormAllocatorInTemp in fused_batch_norm_op.cc. Can you consolidate them to a single location instead of duplicating code, please?

rmlarsen commented on tensorflow/core/ops/ctc_ops.cc:
```diff
@@ -62,6 +62,43 @@ REGISTER_OP("CTCLoss")
     });
 
+REGISTER_OP("CTCLossV2")
+    .Input("inputs: float")
```
Does the CuDNN implementation support types other than float? If so, we should also support them here.

https://github.com/tensorflow/tensorflow/pull/31164 added support for double for CTCLossOp, for example.

houtoms replied:
No, CuDNN only support the float CTCLoss.

sanjoy commented:
Adding Tim Shen to review the stream executor bits.

alextp commented on tensorflow/core/api_def/base_api/api_def_CTCLossV2.pbtxt:
```diff
@@ -0,0 +1,71 @@
+the gradient.  This class performs the softmax operation for you, so inputs
+should be e.g. linear projections of outputs by an LSTM.
+END
+}
```
This should be visibility:hidden so it doesn't show up as tf.ctc_loss_v2

alextp commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -408,6 +408,13 @@ struct PersistentRnnPlanDeleter {
     CHECK_CUDNN_OK(cudnnDestroyPersistentRNNPlan(plan));
   }
 };
+#if CUDNN_VERSION >= 7601
```
Can we not change stream_executor and instead call cudnn directly from the kernel?

houtoms replied:
But I think all the cudnn calls are inside this cuda_dnn.cc. I haven't seen any example to call the cudnn directly from the kernel.

alextp reviewed (changes requested):
This should replace the existing ctc loss for GPUs, right?

kaixih committed:
Put the reusable class CudnnAllocatorInTemp to a separate file

houtoms commented:
@rmlarsen Tx for the commends. I have made several changes. PTAL.

timshen91 commented on tensorflow/stream_executor/dnn.h:
```diff
@@ -190,6 +190,15 @@ class RnnDescriptor {
+//
+// The user is responsible for releasing this descriptor when it is no longer
+// in use. The destructor releases the underlying descriptors.
+class CtcLossDescriptor {
```
Instead of having a base class, can we simply have:
```
class CtcLossDescriptor {
 public:
  explicit CtcLossDescriptor(DataType);
  DataType GetComputeType() const;

 private:
  DataType comp_type_;
};
```

houtoms replied:
I cannot follow the point here. I am following the style of the above RnnDescriptor example. My understanding is we have a device-agnostic base class XDescriptor in dnn.h, then we have specific CuDNNXDescriptor inheriting it in cuda_dnn.h/cc. I can see many other descriptors like RnnDescriptor or RnnSequenceTensorDescriptor did that.

timshen91 replied:
This file is not consistent in how to allow implementations customizing the data structure. For example, CovolutionDescriptor isn't virtual. I think we can start with a simple class, and turn it into virtual later.

timshen91 replied:
> This file is not consistent with itself in how to allow implementations customizing the data structure. For example, CovolutionDescriptor isn't virtual. I think we can start with a simple class, and turn it into virtual later.

timshen91 commented on tensorflow/stream_executor/dnn.h:
```diff
@@ -2383,6 +2402,40 @@ class DnnSupport {
+  //    workspace memory used by this operation. The caller is responsible for
+  //    keeping the memory alive long enough for this operation, and recylces
+  //    afterwards.
+  virtual bool DoCtcLoss(Stream* stream,
```
Can we have something like:
```
virtual bool DoCtcLoss(
  Stream* stream,
  dnn::DataType element_type,
  const dnn::RnnStateTensorDescriptor &probs_desc,
  DeviceMemoryBase probs_data,
  absl::Span<const int32> labels_data,
  absl::Span<const int32> labels_lengths_data,
  absl::Span<const int32> input_lengths_data,
  DeviceMemoryBase costs_data,
  const dnn::RnnStateTensorDescriptor &grads_desc,
  DeviceMemoryBase grads_data,
  const dnn::CtcLossDescriptor &ctc_loss_desc,
  DeviceMemoryBase scratch_memory);
```
? A few points:
* Pass DeviceMemory and Span by value. They are already pointer types, therefore no need to pass by references.
* Generalize on element type, so that one function can support multiple element types, instead of just float. See PrepareForConvolution, the virtual DoConvolve, and the class comment of DnnSupport.
* Separate scratch memory allocation from this function. Users should allocate the scratch memory separately. This is also consistent with cuDNN's API.

houtoms replied:
For the scratch memory allocation, I follow the way used in DoBatchNormalizationForward/Backward and DoRNNForward/Backward. Shouldn't it be consistent with them?
Do I need to generalize the element type in this PR, considering the current cuDNN only support float?

timshen91 replied:
> Please refer to the comment for class DnnSupport, and PrepareForConvolution.

The file itself is already inconsistent. It was way too verbose and I simplified Convolution-related APIs. In the spirit of being as thin of a wrapper as possible (aka simplifying DnnSupport itself), we want to be close to cuDNN / MIOpen without unnecessary cleverness. Please also refer to the comments for DnnSupport class.

> Do I need to generalize the element type in this PR, considering the current cuDNN only support float?

I'd argue it's easier to implement and maintain by adding the element type. The key point is that the element type is just passed down to cuDNN without being much inspected by StreamExecutor.

For example, StreamExecutor doesn't have to know which types are supported vs which are not (because that may change with a different version of cuDNN). If a type is not supported, cuDNN will generate an error, and StreamExecutor just forwards it transparently.

guptapriya commented:
Can you address this comment from before? 
> This should replace the existing ctc loss for GPUs, right?

houtoms commented:
> Can you address this comment from before?
> 
> > This should replace the existing ctc loss for GPUs, right?

Yes, if the environment variable TF_CUDNN_CTC_LOSS is defined. The reason why I cannot make it as the default backend is that CPU and GPU require different data transpose before calling the CTC compute.

alextp commented:
Isn't this similar to how we deal with convolutions? Should we have some
type of layout argument / graph rewrite pass to address this?

I worry this is a regression in usability as it stands...

On Wed, Nov 13, 2019 at 1:36 PM Kaixi Hou <notifications@github.com> wrote:

> Can you address this comment from before?
>
> This should replace the existing ctc loss for GPUs, right?
>
> Yes, if the environment variable TF_CUDNN_CTC_LOSS is defined. The reason
> why I cannot make it as the default backend is that CPU and GPU require
> different data transpose before calling the CTC compute.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/32302?email_source=notifications&email_token=AAABHROJOKBAXPKVSIV4WRTQTRXPVA5CNFSM4IUN3CM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOED7XHNY#issuecomment-553612215>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRNXOYCI6FJBGBNPKSDQTRXPVANCNFSM4IUN3CMQ>
> .
>


-- 
 - Alex

kaixih committed:
Simplified the CtcLossDescriptor

kaixih committed:
Added ElementType and DeviceMemoryBase for CTC Loss

kaixih committed:
Formatting

houtoms commented:
@timshen91 Thanks for your clarification. I followed the convolution example and further simplified the code. PTAL.
Thx.

kaixih committed:
Modified the macros to ompile with old cudnn version

timshen91 commented on tensorflow/stream_executor/dnn.cc:
```diff
@@ -505,6 +505,12 @@ string ConvolutionDescriptor::ToShortString() const {
 
+// -- CtcLossDescriptor
+//
+CtcLossDescriptor::CtcLossDescriptor() {}
```
These can be defined in the header (or left omitted).

timshen91 commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -1969,6 +2049,51 @@ port::Status CudnnSupport::DoRnnBackwardImpl(
+  auto cudnn = cudnn_->GetHandle(parent_, stream);
+
+  SE_ASSIGN_OR_RETURN(DeviceMemory<uint8> workspace,
+                      CreateCtcLossWorkspace(stream, cudnn, ctc_loss_desc,
```
Can you surface this to `ThenCtcLoss`? So that DnnSupport's DoCtcLoss takes a workspace memory, instead of a workspace allocator.

timshen91 commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -1656,6 +1703,39 @@ port::StatusOr<DeviceMemory<uint8>> CreateBatchNormBackwardWorkspace(
+    const CudnnCtcLossDescriptor& ctc_loss_desc,
+    const CudnnRnnStateTensorDescriptor& probs_desc,
+    const CudnnRnnStateTensorDescriptor& grads_desc,
+    const absl::Span<const int32>& labels_data,
```
All `const absl::Span<const int32>&` should be `absl::Span<const int>`. Span acts like a pointer, so there is no need to pass by reference.

kaixih committed:
Move the CtcLossDescriptor constructor/destructor back to the header

Surface the scratch memory allocation to the ThenCtcLoss()

Use the absl::Span as a pointer

houtoms commented:
@timshen91 Thx for your comments. More changes are made accordingly. PTAL.

timshen91 reviewed (dismissed):
LGTM for `stream_executor/*` files. Please wait for other reviewers.

Thanks for the hard work!

alextp commented:
Why do we  need an environment variable? Why can't we treat this the same way as we treat cudnn lstm (using the implementation selector API)? https://github.com/tensorflow/tensorflow/blob/c293fc8040144dd0efbc0f70e483550125fb146e/tensorflow/python/keras/layers/recurrent_v2.py#L1447

Otherwise can't we push the extra transpose and whatnot inside the TF opkernel before dispatching to cudnn?

houtoms commented:
@alextp 
 >> Why do we need an environment variable?
The CPU/GPU preprocessing for the CTC loss is different. So I proposed the environment variable to let users to decide which one to use.
>> Why can't we treat this the same way as we treat cudnn lstm (using the implementation selector API)?
Yes, I am also thinking of using the implementation selector you pointed to. We can define sth like:
```
func1 = preprocessing1+CPU_CTC_loss 
func2 = preprocessing2 + GPU_CTC_loss
call func1 and register func2
``` @qlzh727 Please correct me if I am wrong.

sanjoy commented on tensorflow/core/kernels/ctc_loss_op.cc:
```diff
@@ -186,4 +220,152 @@ REGISTER_CPU(double);
+                                       .HostMemory("labels_indices")
+                                       .HostMemory("labels_values")
+                                       .HostMemory("sequence_length"),
+                        CTCLossOpGPU);
```
Should we register the kernel even if cuDNN is older than 7.6.3 (and the kernel is guaranteed to fail)?

sanjoy commented on tensorflow/core/kernels/ctc_loss_op.cc:
```diff
@@ -186,4 +220,152 @@ REGISTER_CPU(double);
+    const int64 num_classes_raw = inputs_shape.dim_size(2);
+    OP_REQUIRES(
+        ctx, FastBoundsCheck(num_classes_raw, std::numeric_limits<int>::max()),
+        errors::InvalidArgument("num_classes cannot exceed max int"));
```
Why are we not bounds checking `max_time_raw` and `batch_size_raw` also?

sanjoy commented on tensorflow/core/kernels/ctc_loss_op.cc:
```diff
@@ -186,4 +220,152 @@ REGISTER_CPU(double);
+                                        "ignore_longer_outputs_than_inputs_ to"
+                                        "be false"));
+
+    // Convert the labels_indices to labels_lengths
```
End comments with period.  https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

sanjoy commented on tensorflow/core/kernels/ctc_loss_op.cc:
```diff
@@ -186,4 +220,152 @@ REGISTER_CPU(double);
+
+    se::dnn::CtcLossDescriptor ctc_loss_desc;
+    std::unique_ptr<RnnStateTensorDescriptor> probs_desc;
+    std::unique_ptr<RnnStateTensorDescriptor> grads_desc;
```
These should be declared where they are initialized.  https://google.github.io/styleguide/cppguide.html#Local_Variables

sanjoy commented on tensorflow/core/kernels/ctc_loss_op.cc:
```diff
@@ -186,4 +220,152 @@ REGISTER_CPU(double);
+            ->ThenCtcLoss(
+                *probs_desc, probs_data, labels_data, labels_lengths_data,
+                input_lengths_data, &costs_data, *grads_desc, &grads_data,
+                ctc_loss_desc, &workspace_allocator)
```
I may be missing something, but `ctc_loss_desc` is not written to or modified.  If that is correct, it would be nicer to move the declaration immediately before the `ThenCtcLoss` call to make this obvious.

sanjoy replied:
Can you please add a comment stating that the default constructed `ctc_loss_desc` instance is intentional?

sanjoy commented on tensorflow/core/kernels/ctc_loss_op.cc:
```diff
@@ -186,4 +220,152 @@ REGISTER_CPU(double);
+    OP_REQUIRES(ctx, ignore_longer_outputs_than_inputs_ == false,
+                errors::InvalidArgument("GPU CTCLossOp requires "
+                                        "ignore_longer_outputs_than_inputs_ to"
+                                        "be false"));
```
Please prefer `!a` over `a == false` and `a` over `a == true`.

sanjoy commented on tensorflow/core/kernels/ctc_loss_op.cc:
```diff
@@ -186,4 +220,152 @@ REGISTER_CPU(double);
+    OP_REQUIRES_OK(ctx,
+                   ctx->GetAttr("ctc_merge_repeated", &ctc_merge_repeated_));
+    OP_REQUIRES_OK(ctx, ctx->GetAttr("ignore_longer_outputs_than_inputs",
+                                     &ignore_longer_outputs_than_inputs_));
```
Since these are always required to be false why not move the 

```
    OP_REQUIRES(ctx, preprocess_collapse_repeated_ == false,
                errors::InvalidArgument("GPU CTCLossOp requires "
                                        "preprocess_collapse_repeated to be "
                                        "false"));
```

etc. checks here and remove the variables in the class?

sanjoy commented on tensorflow/core/kernels/ctc_loss_op.cc:
```diff
@@ -186,4 +220,152 @@ REGISTER_CPU(double);
 
+#if GOOGLE_CUDA
+class CTCLossOpGPU : public OpKernel {
+
```
Stray empty line.

kaixih committed:
Add impl selector to remove the env var about the CUDNN CTC Loss

kaixih committed:
Use DnnScratchAllocator

kaixih committed:
Variables init and decl in one line; check attr in constructor; check bounds before converting int64 to int; and other minor changes

kaixih committed:
Remove empty lines

houtoms commented:
@sanjoy Thx for the review. More changes are made based on your comments. PTAL.

>> Should we register the kernel even if cuDNN is older than 7.6.3 (and the kernel is guaranteed to fail)?

If users explicitly set the env var TF_CUDNN_CTC_LOSS and their cuDNN is < 7.6.3, it will error out from the stream exector during runtime (as we did for the RNN: if users set the variable sequence length params and their cuDNN < 7.2.1, we will error out. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cuda_dnn.cc#L1729).

sanjoy commented:
> @sanjoy Thx for the review. More changes are made based on your comments. PTAL.
> 
> > > Should we register the kernel even if cuDNN is older than 7.6.3 (and the kernel is guaranteed to fail)?
> 
> If users explicitly set the env var TF_CUDNN_CTC_LOSS and their cuDNN is < 7.6.3, it will error out from the stream exector during runtime (as we did for the RNN: if users set the variable sequence length params and their cuDNN < 7.2.1, we will error out. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cuda_dnn.cc#L1729).

Let me rephrase: what is the use case for even registering a kernel that will always fail during runtime?  Since that's what will happen if cuDNN is older than 7.6.3 right?

houtoms commented:
Do you think we need to add cudnn version checking macros to protect this?

On Thu, Dec 5, 2019 at 12:25 PM Sanjoy Das <notifications@github.com> wrote:

> @sanjoy <https://github.com/sanjoy> Thx for the review. More changes are
> made based on your comments. PTAL.
>
> Should we register the kernel even if cuDNN is older than 7.6.3 (and the
> kernel is guaranteed to fail)?
>
> If users explicitly set the env var TF_CUDNN_CTC_LOSS and their cuDNN is <
> 7.6.3, it will error out from the stream exector during runtime (as we did
> for the RNN: if users set the variable sequence length params and their
> cuDNN < 7.2.1, we will error out.
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cuda_dnn.cc#L1729
> ).
>
> Let me rephrase: what is the use case for even registering a kernel that
> will always fail during runtime? Since that's what will happen if cuDNN is
> older than 7.6.3 right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/32302?email_source=notifications&email_token=AA6Q5EAZ3DAU7Y7JI6QQEO3QXFPUNA5CNFSM4IUN3CM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGCA5NA#issuecomment-562302644>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA6Q5EAPDXFS4XIIWKALALTQXFPUNANCNFSM4IUN3CMQ>
> .
>

sanjoy commented:
> Do you think we need to add cudnn version checking macros to protect this?

Yeah, I'm asking why not do:

```
#if GOOGLE_CUDA && CUDNN_VERSION >= 7603
class CTCLossOpGPU : public OpKernel {
 public:
  explicit CTCLossOpGPU(OpKernelConstruction* ctx) : OpKernel(ctx) {
    bool preprocess_collapse_repeated_;
    bool ctc_merge_repeated_;
    bool ignore_longer_outputs_than_inputs_;
    ...

 private:
  TF_DISALLOW_COPY_AND_ASSIGN(CTCLossOpGPU);
};

REGISTER_KERNEL_BUILDER(Name("CTCLossV2").Device(DEVICE_GPU)
                                       .HostMemory("labels_indices")
                                       .HostMemory("labels_values")
                                       .HostMemory("sequence_length"),
                        CTCLossOpGPU);
#endif
```

Btw, this is not a rhetorical question.  If you have a use case for always registering the kernel then I'm happy as well.

houtoms commented:
> If you have a use case for always registering the kernel then I'm happy as well.

Yeah, we could do as you suggested. But I think it would bring more problems. What about the `REGISTER_OP("CTCLossV2")` in `tensorflow/core/ops/ctc_ops.cc`. And what happens in python when we call `gen_ctc_ops.ctc_loss_v2`. Do we also need to protect them? It sounds like every time when we have some new op, we need to protect all the related code from front end python API, kernel definition, all the way to the stream executor. (I think that is why in the RNN we only have the macro protect code only in stream executor.)  

Also, I have never seen any example to protect the register-op codes using CUDNN_VERSION in the tensorflow/core/kernels (Please correct me if I am wrong.).

sanjoy commented:
> > If you have a use case for always registering the kernel then I'm happy as well.
> 
> Yeah, we could do as you suggested. But I think it would bring more problems. What about the `REGISTER_OP("CTCLossV2")` in `tensorflow/core/ops/ctc_ops.cc`.

I don't think we should guard the **op** because we could (in the future) have a CPU or CUDA kernel for it that does not depend on cuDNN.  I agree that piping in the cuDNN version check all the way up to Python would cumbersome.

The problem I have with the current formulation is that we're pretending that we have a GPU implementation for cuDNN < 7.6.3, even though we really don't have an implementation.

kaixih committed:
Avoid to register the ctc loss kernel when cudnn is older than 7.6.3

houtoms commented:
@sanjoy Done. PTAL.

sanjoy commented on tensorflow/stream_executor/dnn.h:
```diff
@@ -190,6 +190,13 @@ class RnnDescriptor {
 };
 
+// Describes a CTC loss operation.
+class CtcLossDescriptor {
```
I'm missing something here: why is this class empty (please add comments with the rationale).

sanjoy replied:
Maybe I'm misusing the github UI, but I still see is empty `CtcLossDescriptor` class.

kaixih committed:
Remove the empty CtcLossDescriptor

houtoms commented:
@sanjoy Yes, I went though the code again and the empty class of CtcLossDescriptor in dnn.h seems to be unnecessary. Removed it. PTAL.

gbaned commented:
@houtoms Can you please resolve conflicts? Thanks!

kaixih committed:
Solved a conflict

houtoms commented:
The conflict is solved. PTAL. Thx.

sanjoy reviewed (changes requested):
Other than the minor comment the non-Python code looks good to me.  If Alex and Rasmus are OK with the API then this is good to go.

/CC @alextp @rmlarsen

alextp commented on tensorflow/python/ops/ctc_ops.py:
```diff
@@ -656,26 +698,36 @@ def ctc_loss_v2(labels,
       raise ValueError(
           "blank_index must be given when using SparseTensor labels.")
 
+    _ctc_use_cudnn = os.environ.get("TF_CUDNN_CTC_LOSS", "0")
```
Environment variables are not ok for configuring TF.

We suggested earlier using the same mechanism we use for cudnn lstm.

houtoms replied:
Hi @alextp , I made an implementation selector version internally but it didn't work. My guess is the mechanism might not work well with the loss function. Contacted with @qlzh727 (the owner of implementation selector) already.

kaixih committed:
remove empty CtcLossDescriptor in dnn.h

houtoms commented:
@sanjoy Thanks for pointing out. Removed that empty class. PTAL.

sanjoy commented:
> Other than the minor comment the non-Python code looks good to me. If Alex and Rasmus are OK with the API then this is good to go.
> 
> /CC @alextp @rmlarsen

This still applies.

houtoms commented:
@sanjoy @alextp  , I have replaced the previous environment variable with the implementation selector (Thx @qlzh727 for helping me out with some test cases.) Now, we don't need the env var to control if cuDNN is used or not. The runtime can automatically determine that if GPU is available or not.

I added another python function to contain this new implement (ie. ctc_loss_v3), which is only available in TF2.

Please help me find the reviewers to review this part. Thx.

alextp commented on tensorflow/python/ops/ctc_ops.py:
```diff
@@ -42,6 +46,28 @@
 from tensorflow.python.util import nest
 from tensorflow.python.util.tf_export import tf_export
 
+import os
```
The linter will complain; standard python imports need to be above all others

houtoms replied:
Sure. Removed this unused import.

kaixih committed:
remove unused import

alextp commented on tensorflow/core/api_def/base_api/api_def_CTCLossV2.pbtxt:
```diff
@@ -0,0 +1,71 @@
+op {
+  graph_op_name: "CTCLossV2"
```
Can you add a line here saying "visibility: HIDDEN"; this will prevent the generation of a tf.ctc_loss_v2 API

houtoms replied:
Sure. Done.

alextp reviewed (changes requested):
Looks good! Just a couple of minor nits and we can approve.

Thanks!

kaixih committed:
Set CTCLossV2 to visibility:HIDDEN

alextp commented on tensorflow/tools/api/golden/v2/tensorflow.pbtxt:
```diff
@@ -576,6 +576,10 @@ tf_module {
     name: "cosh"
```
Now you can revert this file to make the API tests pass again

alextp commented on tensorflow/tools/api/golden/v1/tensorflow.pbtxt:
```diff
@@ -1068,6 +1068,10 @@ tf_module {
     name: "cross"
```
Now you can revert this file to make the API tests pass again

houtoms replied:
Done. PTAL.

kaixih committed:
update goldens

houtoms commented:
Thx for the update.
